compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -3356,3 +3438,93 @@ fn add_lld_args(
         }
     }
 }
+
+/// Create an ELF .so stub file for raw-dylib.
+/// It exports all the provided symbols, but is otherwise empty.
+fn create_elf_raw_dylib_stub(symbols: &[DllImport]) -> Vec<u8> {
+    use object::elf;
+    use object::write::elf as write;
+
+    let mut stub_buf = Vec::new();
+
+    // When using the low-level object::write::elf, the order of the reservations
+    // needs to match the order of the writing.
+
+    let mut stub = write::Writer::new(object::Endianness::Little, true, &mut stub_buf);
+
+    // These initial reservations don't reserve any space yet.
+    stub.reserve_null_dynamic_symbol_index();
+
+    let dynstrs = symbols
+        .iter()
+        .map(|sym| {
+            stub.reserve_dynamic_symbol_index();
+            (sym, stub.add_dynamic_string(sym.name.as_str().as_bytes()))
+        })
+        .collect::<Vec<_>>();
+
+    stub.reserve_shstrtab_section_index();
+    let text_section_name = stub.add_section_name(".text".as_bytes());
+    let text_section = stub.reserve_section_index();
+    stub.reserve_dynstr_section_index();
+    stub.reserve_dynsym_section_index();
+
+    // These reservations determine the actual layout order of the object file.
+    stub.reserve_file_header();
+    stub.reserve_shstrtab();
+    stub.reserve_section_headers();
+    stub.reserve_dynstr();
+    stub.reserve_dynsym();
+
+    // File header
+    stub.write_file_header(&write::FileHeader {
+        os_abi: object::elf::ELFOSABI_SYSV,
+        abi_version: 0,
+        e_type: object::elf::ET_DYN,
+        e_machine: object::elf::EM_X86_64,
+        e_entry: 0,
+        e_flags: 0,
+    })
+    .unwrap();
+
+    // .shstrtab
+    stub.write_shstrtab();
+
+    // Section headers
+    stub.write_null_section_header();
+    stub.write_shstrtab_section_header();
+    // Create a dummy .text section for our dummy symbols.
+    stub.write_section_header(&write::SectionHeader {
+        name: Some(text_section_name),
+        sh_type: elf::SHT_PROGBITS,
+        sh_flags: 0,
+        sh_addr: 0,
+        sh_offset: 0,
+        sh_size: 0,
+        sh_link: 0,
+        sh_info: 0,
+        sh_addralign: 1,
+        sh_entsize: 0,
+    });
+    stub.write_dynstr_section_header(0);
+    stub.write_dynsym_section_header(0, 1);
+
+    // .dynstr
+    stub.write_dynstr();
+
+    // .dynsym
+    stub.write_null_dynamic_symbol();
+    for (_, name) in dynstrs {
+        stub.write_dynamic_symbol(&write::Sym {
+            name: Some(name),
+            st_info: (elf::STB_GLOBAL << 4) | elf::STT_NOTYPE,
+            st_other: elf::STV_DEFAULT,
+            section: Some(text_section),
```

Comment:
Is it actually necessary to add those symbols to sections?

Reply:
Yes, this appear to be necessary. If I change it to `None`, I get a linker error (I guess `None`/0 means undefined?).

Reply:
Not providing a section means it's either undefined or an absolute value, I don't know how the object crate converts `None` here.